contrib: Add ContainerLab-based BGP CPlane development environment

YutaroHayakawa opened the pull request:
Please see the commit message for more details.

```release-note
contrib: Add ContainerLab-based BGP CPlane development environment
```

YutaroHayakawa committed:
contrib: Add ContainerLab-based BGP CPlane development environment

It is not always easy for developers to prepare an environment that
easily deploy and test BGP Control Plane since it requires unusual
network topology to make BGP peering and test reachability from external
node that doesn't run Cilium.

This commit adds BGP Control Plane development environments based on
ContainerLab and Kind with IPv4/IPv6 single-stack and dual-stack. The
topology looks like below. Note that server0-2 are directly connected
to router0 with veth.

                                    |---server0(Cilium, ASN: 65001)
                                    |
   router0(FRRouting, ASN: 65000)---|---server1(Cilium, ASN: 65002)
                                    |
                                    |---server2(netshoot, No BGP)

router0 has FRR on it. You can use `docker exec -it <container name>
vtysh` to do any manual configuration. server0 and server1 are netshoot
container that shares network namespace with Kind node. You can use
netshoot container through `docker exec` (useful for
debugging/troubleshooting) and use Cilium through `kubectl exec`.
server2 is a netshoot container you can use for testing external
connectivity.

The development workflow looks like following.

1. Deploy lab

$ make kind-bgp-<v4|v6|dual>

2. Install Cilium with your favorite tool. Mandatory Helm values are
   in contrib/containerlab/bgp-cplane-dev-<v4|v6|dual>/values.yaml.

$ cilium install -f contrib/containerlab/bgp-cplane-dev-<v4|v6|dual>/values.yaml

3. Do basic peering

$ make kind-bgp-<v4|v6|dual>-apply-policy

4. Validate peering state

$ cilium bgp peers

5. Enjoy development

6. Cleanup lab

$ make kind-bgp-<v4|v6|dual>-down

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>
Co-authored-by: Ryan Drew <ryan.drew@isovalent.com>
Co-authored-by: Daneyon Hansen <daneyon.hansen@solo.io>

rastislavs reviewed (approved):
nice!

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,83 @@
+Development Environment
+=======================
+
+BGP Control Plane requires BGP peer for testing. You can use `ContainerLab <https://containerlab.dev/>`_ and `Kind <https://kind.sigs.k8s.io/docs/user/configuration/>`_-based development environment. The following diagram shows the topology.
```
Couple of suggestions here for grammar and word choice.
```suggestion
BGP Control Plane requires a BGP peer for testing. This section describes a `ContainerLab <https://containerlab.dev/>`_ and `Kind <https://kind.sigs.k8s.io/docs/user/configuration/>`_-based development environment. The following diagram shows the topology:
```

learnitall replied:
Additionally, can you modify these links so they are block-level hyperlink targets? See https://docs.cilium.io/en/stable/contributing/docs/docsstyle/#links

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,83 @@
+.. image:: _static/bgp-cplane-dev-topo.png
+    :align: center
+
+Role of each nodes are following
```
I'm not sure if the grammar is right here, I left a suggestion for rewording this.

```suggestion
The following describes the role of each node:
```

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,83 @@
+
+Role of each nodes are following
+
+* ``router0`` is an `FRRouting (FRR) <https://frrouting.org/>`_ router. It is pre-configured with a minimal peering settings with server0 and server1.
```
The "a" tells the reader that the noun it applies to is singular, but "settings" is plural. I'd suggest the following:

```suggestion
* ``router0`` is an `FRRouting (FRR) <https://frrouting.org/>`_ router. It is pre-configured with minimal peering settings with server0 and server1.
```

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,83 @@
+Role of each nodes are following
+
+* ``router0`` is an `FRRouting (FRR) <https://frrouting.org/>`_ router. It is pre-configured with a minimal peering settings with server0 and server1.
+* ``server0`` and ``server1`` are Kind nodes (technically, they are ``nicolaka/netshoot`` containers that share network namespaces with Kind nodes).
```
Do you think you could reword this so there aren't parentheses used here? This is really cool that these are netshoot containers that share a netns with kind nodes, it shows off the power and usefulness of containerlab. I think it makes sense to highlight that. Here's an example of what I'm thinking:

```suggestion
* ``server0`` and ``server1`` are ``nicolaka/netshoot`` containers that each share a network namespace with their own Kind node.
```

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,83 @@
+Install Cilium on the Lab
+-------------------------
+
+Install Cilium on the lab with your favorite way. Following example assumes you are modifying the source and want to build your image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. You can add Helm values to deploy BGP Control Plane with different Cilium configuration.
```
There are a couple of changes that I'd recommend here for grammar and word choice:

```suggestion
Install Cilium on the lab with your favorite way. The following example assumes you are modifying the source and want to build your own image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. If needed, you can add Helm values to deploy BGP Control Plane with a different Cilium configuration.
```

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,83 @@
+Peering with Router
+-------------------
+
+Peering Cilium nodes with FRR by applying CiliumBGPPeeringPolicy.
```
A couple of grammar suggestions to reword this as an instruction for the user:

```suggestion
Peer Cilium nodes with FRR by applying a CiliumBGPPeeringPolicy:
```

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,83 @@
+.. code-block:: shell-session
+
+   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.1.2
+   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.2.2
```
Is there any output here that the user should expect if the connectivity is successful? Or is there no output and a zero RC?

YutaroHayakawa replied:
Good point. Updated with the expected output and description for it.

learnitall reviewed (changes requested):
Thank you for your patience. This is a really cool doc! I just have a couple of changes that I'd recommend for grammar and structure and then we can get this merged in.

danehans commented on contrib/containerlab/bgp-cplane-dev-dual/.gitignore:
```diff
@@ -0,0 +1,2 @@
+clab-bgp-cplane-dev-dual/*
+.topo.yaml.bak
```
I don't see `.topo.yaml.bak` used anywhere other than being referenced by the gitignores. Is this an accidental leftover from testing?

Why ignore all files in this directory?

YutaroHayakawa replied:
These are the file/dir automatically generated by ContainerLab. `clab-bgp-cplane-dev-dual` is a directory to store the runtime state and .topo.yaml.bak is a backup manifest generated when you create the lab.

danehans replied:
Thanks for the input. Might not be a bad idea to document ^ somewhere.

YutaroHayakawa replied:
Good idea, documented in the comments on the `.gitignore` files.

danehans commented on contrib/containerlab/bgp-cplane-dev-dual/topo.yaml:
```diff
@@ -0,0 +1,74 @@
+      exec:
+      - ip addr add 10.0.2.2/24 dev net0
+      - ip addr add fd00:10:0:2::2/64 dev net0
+      - ip route add 10.0.0.0/8 via 10.0.2.1 dev net0
```
L49 applies here too, correct?

YutaroHayakawa replied:
Correct. Do you mean we should have a comment here as well? I omitted it because it is already mentioned on the doc that the roles of server0 and server1 are the same, and assumed that people could understand what does mean from server0's configuration.

danehans replied:
I don't think adding the same comment throughout is necessary but I don't think it hurts either... your call.

YutaroHayakawa replied:
Yep, nothing hurts. I added comments.

danehans commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,99 @@
+BGP Control Plane
+=================
+
+This section is specific to BGP Control Plane contributions.
```
nit: consider linking `BGP Control Plane` to the BGP CP [docs](https://docs.cilium.io/en/stable/network/bgp-control-plane/).

YutaroHayakawa replied:
Good idea. Done.

danehans commented on contrib/containerlab/bgp-cplane-dev-dual/Makefile:
```diff
@@ -0,0 +1,10 @@
+deploy:
+	kind create cluster --config cluster.yaml
+	sudo containerlab -t topo.yaml deploy
```
I already had the `containerlab` bin installed so this target worked fine for me. Should a note be added to the doc that states this requirement and the need for sudo privs? If so, you may also want to include a specific version of the bin.

YutaroHayakawa replied:
Indeed. I added a new `Prerequisites` section.

danehans commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,99 @@
+.. only:: not (epub or latex or html)
+
```
Adding a comment here since I can't comment on the png file. Consider adding the Kind node name to the diagram, e.g. `server0 (Kind control-plane Node, Cilium)`.

YutaroHayakawa replied:
Good idea! I added the control-plane and worker, respectively.

danehans commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,99 @@
+
+.. code-block:: shell-session
+
+   $ cilium bgp peers
```
Consider adding the expected response from the command or add a short note that states the expected session state, routers received/advertised, etc.

learnitall replied:
+1 from me, that's a great idea.

YutaroHayakawa replied:
Thanks, I added an expected result and description.

danehans commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,99 @@
+
+   $ make docker-cilium-image docker-operator-generic-image
+   $ kind --name bgp-cplane-dev-v4 load docker-image quay.io/cilium/cilium:latest quay.io/cilium/operator-generic:latest
+   $ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml --helm-set image.override="quay.io/cilium/cilium:latest" --helm-set image.pullPolicy=Never --helm-set operator.image.override="quay.io/cilium/operator-generic:latest" --helm-set operator.image.pullPolicy=Never
```
This workflow could be simplified by adding the ability to specify the kind cluster name. For example:

```sh
$ git diff Makefile
diff --git a/Makefile b/Makefile
index 3730ee6c01..bd0d692cb0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,8 @@
 # Copyright Authors of Cilium
 # SPDX-License-Identifier: Apache-2.0

+CLUSTER_NAME ?= kind
+
 ##@ Default
 all: precheck build postcheck ## Default make target that perform precheck -> build -> postcheck
        @echo "Build finished."
@@ -576,7 +578,7 @@ kind-build-image-agent: ## Build cilium-dev docker image

 $(eval $(call KIND_ENV,kind-image-agent))
 kind-image-agent: kind-ready kind-build-image-agent ## Build cilium-dev docker image and import it into kind.
-       $(QUIET)kind load docker-image $(LOCAL_AGENT_IMAGE)
+       $(QUIET)kind load docker-image $(LOCAL_AGENT_IMAGE) -n $(CLUSTER_NAME)

 $(eval $(call KIND_ENV,kind-build-image-operator))
 kind-build-image-operator: ## Build cilium-operator-dev docker image
@@ -584,7 +586,7 @@ kind-build-image-operator: ## Build cilium-operator-dev docker image

 $(eval $(call KIND_ENV,kind-image-operator))
 kind-image-operator: kind-ready kind-build-image-operator ## Build cilium-operator-dev docker image and import it into kind.
-       $(QUIET)kind load docker-image $(LOCAL_OPERATOR_IMAGE)
+       $(QUIET)kind load docker-image $(LOCAL_OPERATOR_IMAGE) -n $(CLUSTER_NAME)

 $(eval $(call KIND_ENV,kind-build-clustermesh-apiserver))
 kind-build-clustermesh-apiserver: ## Build cilium-clustermesh-apiserver docker image
```

A user would then:

```sh
$ CLUSTER_NAME=bgp-cplane-dev-v4 make kind-image
$ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml
```

YutaroHayakawa replied:
Following the workflow of the `make kind-xxx` target sounds good to me. I put your suggestion into a separate commit and updated the document. Please take a look.

danehans reviewed (commented):
I took this PR for a test drive and this will be a nice way to test/dev/demo BGP CP.

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,99 @@
+   $ make kind-bgp-v4
+
+.. note::
+        Above example setups IPv4 single-stack environment. You can change ``v4`` part to ``v6`` or ``dual`` to setup IPv6 single-stack or dual-stack environment respectively. Same for all below examples.
```
Can you replace the use of "Above" and "below" here? Sometimes translators have a hard time with these words in this context. Using "prior" and "following" are great alternatives.

Additionally, because "setup" is used as a verb here, it needs to be separated into two words. It's something that I had to learn when reviewing this :). This page is really helpful: https://www.grammarbook.com/blog/spelling/setup-vs-set-up/.

I also left some suggestions for grammar.

```suggestion
        The prior example sets up an IPv4 single-stack environment. You can change the ``v4`` part to ``v6`` or ``dual`` to set up an IPv6 single-stack or dual-stack environment respectively. The same goes for the following examples.
```

YutaroHayakawa replied:
Thanks for the input! I didn't know they were different words 😮 

Applied your suggestion.

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,99 @@
+Validating Connectivity to Cilium Nodes from Non-Cilium Node
+------------------------------------------------------------
+
+The below example validates connectivity from server2 to server0 (10.0.1.2) and server1 (10.0.2.2). You should see the packet goes through router0 (10.0.3.1).
```
I think this needs to be "go" here instead of "goes"

```suggestion
The below example validates connectivity from server2 to server0 (10.0.1.2) and server1 (10.0.2.2). You should see the packet go through router0 (10.0.3.1).
```

YutaroHayakawa replied:
The `packet` needed to be `packets` as well. Fixed.

learnitall commented on Documentation/contributing/development/bgp_cplane.rst:
```diff
@@ -0,0 +1,99 @@
+   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.1.2 -r
+   Start: 2023-10-04T02:24:58+0000
+   HOST: server2                     Loss%   Snt   Last   Avg  Best  Wrst StDev
+   1.|-- 10.0.3.1                   0.0%    10    0.0   0.1   0.0   0.1   0.0
```
I'm not familiar with the expected output here, but is there some whitespace misalignment here? The data in the columns don't seem to line up with their associated headers. For instance, I would have expected:

```
$ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.2.2 -r
Start: 2023-10-04T02:25:58+0000
HOST: server2                     Loss%   Snt   Last   Avg  Best  Wrst StDev
1.|-- 10.0.3.1                     0.0%    10    0.1   0.1   0.0   0.1   0.0
2.|-- 10.0.2.2                     0.0%    10    0.1   0.1   0.0   0.2   0.0
```

The same goes for the table in the second command.

Sometimes weird stuff can happen when copying and pasting command output, so I just wanted to double check.

YutaroHayakawa replied:
Yes, as you said, something wrong happened when I copied and pasted the output. I fixed it.

learnitall reviewed (changes requested):
The changes look great! I have a couple last notes for you in regards to some of the stuff you changed, as well as something that I accidentally missed in my last review.

YutaroHayakawa committed:
Allow specifying Kind cluster name in make kind-image

Allow specifying Kind cluster name with KIND_CLUSTER_NAME environment
variable in kind-image-agent, kind-image-operator, and kind-image. So
that we can load image to the target cluster with the name other than
`kind`.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>
Suggested-by: Daneyon Hansen <daneyon.hansen@solo.io>

learnitall reviewed (approved):
Looks great!
